client: add WhereAny() and let Where() take multiple models

dcbw opened the pull request:
There are cases where we'd like to take a list of models and return a list of results without grabbing the row cache lock for every single item. Batch them up into a Where() condition that can then be List()-ed right after, which is about 2x faster than Get()-ing each item individually.

Before:
```
BenchmarkAPIListMultiple/multiple_results_one_at_a_time_with_Get-8         	     212	   5534355 ns/op
BenchmarkAPIListMultiple/multiple_results_one_at_a_time_with_Get-8         	     212	   5640041 ns/op
BenchmarkAPIListMultiple/multiple_results_one_at_a_time_with_Get-8         	     206	   5680483 ns/op
```
After:
```
BenchmarkAPIListMultiple/multiple_results_in_a_batch_with_WhereAny-8       	     420	   2795212 ns/op
BenchmarkAPIListMultiple/multiple_results_in_a_batch_with_WhereAny-8       	     412	   2895229 ns/op
BenchmarkAPIListMultiple/multiple_results_in_a_batch_with_WhereAny-8       	     412	   2864806 ns/op
```

This PR also adds a WhereAny() call that acts like WhereAll() except with OR semantics, which is equivalent to the previous Where() call including conditions. Thus we now have:

1. Where()  - accepts one or more models and matches on indexes and model data
2. WhereAny() - takes a model and one or more conditions, matching each condition with logical OR
3. WhereAll()  - takes a model and one or more conditions, matching all conditions with a logical AND


@jcaamano @flavio-fernandes @trozet @tssurya

dcbw commented:
> I have a feeling that this `WhereMany` is how `Where` should have been all along. What we could do is have `WhereAny` instead of `WhereMany`, that basically deprecates `Where`.

@jcaamano updated to be WhereAny(). I still need to add testcases.

jcaamano commented on cache/cache.go:
```diff
@@ -166,47 +166,56 @@ func (r *RowCache) Row(uuid string) model.Model {
+			vals := r.indexes[indexSpec.index]
+			if uuids, ok := vals[val]; ok {
+				for uuid := range uuids {
+					results[uuid] = r.rowByUUID(uuid)
```
Would it make sense to check if uuid is already in results to avoid the potentially unnecessary clone done in rowByUUID?

dcbw replied:
Yeah it would.

jcaamano commented on client/client.go:
```diff
@@ -1313,8 +1313,9 @@ func (o *ovsdbClient) List(ctx context.Context, result interface{}) error {
 // Where implements the API interface's Where function
-func (o *ovsdbClient) Where(m model.Model) ConditionalAPI {
-	return o.primaryDB().api.Where(m)
+// DEPRECATED in favor of WhereAny()
```
Not deprecated in the end though?

jcaamano reviewed (changes requested):
@dcbw could you add tests where you pass models of different types, in the same places you added new tests? That should fail in some way, right?

jcaamano commented:
Wan't to run it by ovn-k to discard issues?

> @dcbw could you add tests where you pass models of different types, in the same places you added new tests? That should fail in some way, right?

Just cross-checking you read this. What I see is that you use the first model of the list to know which type it is and get the RowCache matching that type. And then in `RowsByModels` you skip any model that does not match that type. I am considering if an error would be more appropriate, as skipping can lead to the interpretation that the request made sense but the looked after models did not exist.

dcbw commented:
> And then in `RowsByModels` you skip any model that does not match that type.

@jcaamano RowsByModels() doesn't return an error; I could make it return one though.

dcbw commented:
@jcaamano how about now? changed rowsByModel() to return an error and followed that through.

Also added testcases as requested to ensure model lists return an error when the models aren't the same type as the first element.

jcaamano commented on client/condition.go:
```diff
@@ -87,17 +87,31 @@ func (c *equalityConditional) Matches() (map[string]model.Model, error) {
 func (c *equalityConditional) Generate() ([][]ovsdb.Condition, error) {
 	models, _ := c.Matches()
```
I would probably not ignore this error now we have opened the API of `RowsByModels` to return one.

dcbw replied:
@jcaamano I just reworked the commit to respect the errors here and test for them in api_test.go.

jcaamano commented on client/condition.go:
```diff
@@ -87,17 +87,31 @@ func (c *equalityConditional) Matches() (map[string]model.Model, error) {
+func newEqualityConditional(table string, cache *cache.TableCache, models []model.Model) (Conditional, error) {
+	// Ensure all models are the same type
+	expectedType := reflect.TypeOf(models[0])
+	for _, m := range models {
+		if reflect.TypeOf(m) != expectedType {
+			return nil, fmt.Errorf("all models must by of same type %s", expectedType)
+		}
+	}
```
I would not make this check here.

For cache access, Matches()->RowsByModel will already fail.
For server access, we can make this check above in Generate where we already iterate the models.

So we can avoid doing an extra iteration here. A bit of a nit but thinking at scale. Unless there is a reason we need to check early here that I am missing.

jcaamano replied:
I don't think we need anything server side. I did not word correctly what I meant:

Let's say that `Matches` returns `len(models) == 0`. We probably can't assume that it has checked that the models are of the same type in this case (we know it does, but that's an implementation detail that might change). So when this happens I would still check that all models are of the same type. One place we can do this without an extra iteration is in `generateConditionsFromModels` as we are already iterating the models there.

Not a strong opinion on this one, so I will let you decide whether we need this or not.

jcaamano commented on client/api.go:
```diff
@@ -26,14 +26,19 @@ type API interface {
-	// where operations apply to elements that match all the conditions
+	// Create a ConditionalAPI from a Model's index data, where operations
+	// apply to elements that match the values provided in the model.Model
+	// according to the indexes.
```
Should we document the models need to be the same type? I don't know if this would be perceived as a limitation by the user in which case it might be unexpected unless documented.

dcbw replied:
Good suggestion; fixed.

dcbw commented:
@jcaamano ran a 120 density-light test (which is what exercises this most) downstream and got

a7da83bf-node-density-20220823 (P99=4.19s AVG=3.04s)

which is slightly better than usual. Timings show that:

```
I0823 20:08:22.492399       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 234 took router: 134.883µs, nats: 507.978µs); reconcile took 19.239µs
I0823 20:08:22.532599       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 237 took router: 129.662µs, nats: 555.431µs); reconcile took 21.32µs
I0823 20:08:22.587449       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 239 took router: 129.239µs, nats: 562.158µs); reconcile took 20.164µs
I0823 20:08:22.637912       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 236 took router: 165.132µs, nats: 484.924µs); reconcile took 18.919µs
I0823 20:08:22.691104       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 235 took router: 126.207µs, nats: 595.055µs); reconcile took 39.61µs
I0823 20:08:22.739018       1 router.go:897] ####### CreateOrUpdateNATsOps; (getRouterNats 237 took router: 129.993µs, nats: 558.256µs); reconcile took 19.946µs
```

looking up 230+ NAT objects in a batch almost never take more than 1ms while before they could take a *lot* more time than that.

jcaamano commented on client/condition.go:
```diff
@@ -77,27 +77,36 @@ func (c *equalityConditional) Matches() (map[string]model.Model, error) {
 // on the index.
 func (c *equalityConditional) Generate() ([][]ovsdb.Condition, error) {
-	models, _ := c.Matches()
+	models, err := c.Matches()
+	if err != nil {
+		return nil, err
+	}
```
We might need to filter out `ErrNotFound` from this error handling. Sorry I have not warned you before.

We will get `ErrNotFound` from `tableCache := c.cache.Table(c.tableName)` in `Matches` for example when we are not monitoring that table. I think in this case we should behave as if `len(models) == 0`.

dcbw replied:
Fixed.

jcaamano commented:
LGTM
